game-state-service

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python3
+
+from gi.repository import Gio
+from gi.repository import GLib
+from gi.repository import GObject
+
+import dbus
+import dbus.service
+import dbus.mainloop.glib
+import json
+
+DBUS_IFACE = 'com.endlessm.GameStateService'
+DBUS_PATH = '/com/endlessm/GameStateService'
+
+
+class NotFoundException(dbus.DBusException):
+    _dbus_error_name = 'com.endlessm.GameStateService.NotFoundException'
+
+
+class GameState(GObject.Object):
+
+    _MONITOR_RATE = 1000
+
+    def __init__(self):
+        GObject.Object.__init__(self)
+        self._state = {}
+        self._monitor = None
+        self._monitor_changed_id = None
+        self._monitor_restart_id = None
+        self._monitor_start()
+        self._state_from_file()
+
+    def _get_state_file(self):
+        path = GLib.build_filenamev([GLib.get_user_data_dir(), 'state.json'])
+        return Gio.File.new_for_path(path)
+
+    def _monitor_start(self):
+        state_file = self._get_state_file()
+        self._monitor = state_file.monitor_file(
+            Gio.FileMonitorFlags.NONE,
+            None)
+        self._monitor.set_rate_limit(self._MONITOR_RATE)
+        self._monitor_changed_id = self._monitor.connect(
+            'changed',
+            self._monitor_change_cb)
+
+    def _monitor_change_cb(self, monitor, child, other_file, event):
+        if event == Gio.FileMonitorEvent.CHANGES_DONE_HINT:
+            self._state_from_file()
+
+    def _monitor_restart_cb(self):
+        self._monitor_restart_id = None
+        self._monitor.handler_unblock(self._monitor_changed_id)
+        return False
+
+    def _state_from_file(self):
+        state_file = self._get_state_file()
+        try:
+            success, state_data, tag = state_file.load_contents(None)
+            self._state = json.loads(state_data.decode())
+        except:
+            pass
+
+    def _state_to_file(self):
+        state_data = json.dumps(self._state,
+                                sort_keys=True,
+                                indent=2,
+                                separators=(',', ': '))
+        state_file = self._get_state_file()
+
+        # always block the monitor when writing to the state file
+        if self._monitor_restart_id is None:
+            self._monitor.handler_block(self._monitor_changed_id)
+        else:
+            GLib.source_remove(self._monitor_restart_id)
+
+        try:
+            state_file.replace_contents(
+                bytes(state_data.encode()),
+                None,
+                False,
+                0,
+                None)
+        except:
+            pass
+        finally:
+            self._monitor_restart_id = GLib.timeout_add(
```

Comment:
I think there are two other aspects to this pattern:
- we should defer writing the new state file to an idle or a short timeout. That way, if someone calls `set()` in a tight loop, we only write once at the end
- since you use the changes-done hint to reload the state from file, there should be no need to stop the file monitor while you write. But in fact there should be no need to have a file monitor at all, since all the writes go through the DBus service and you already update the state upon a call to `set()`

Reply:
I understood from this morning's discussion that the idea behind the file monitor is that it's easily hackable by editing the text file.

Reply:
For development or is that something that we'd expose to users?

Reply:
In any case for development, and also as something that curious users could discover.

Reply:
> Are you avoiding some kind of race condition?

Yes. Otherwise, it would always catch it owns writes. We don't want that.

> we should defer writing the new state file to an idle or a short timeout

I considered this, but when mixed with the "just edit the file" feature, I though it could turned into a tricky game. But I can give it a try.

> But in fact there should be no need to have a file monitor at all

There is, the "just edit the file feature" is something we want to easily play with scenarios, and interact with the game state.

> For development or is that something that we'd expose to users?

For development really, but like @ptomato said, it would be cool if they find it and hack it.

Reply:
We should make a decision one way or the other on the "curious user" use case before merging this.

Reply:
I don't think that's an interesting use case for now. Let's proceed without and we can revisit later if we want to bring it back.